writers/writers_test.go

```diff
@@ -0,0 +1,209 @@
+	"golang.org/x/exp/maps"
+)
+
+type bCase struct {
+	name string
+	wr   writers.Writer
+	rec  func() arrow.Record
+}
+
+func BenchmarkWriterMemory(b *testing.B) {
+	batchwriterOpts := map[string][]batchwriter.Option{
+		"defaults":           nil,
+		"batch10k bytes100M": {batchwriter.WithBatchSizeBytes(100000000), batchwriter.WithBatchSize(10000)},
+	}
+	mixedbatchwriterOpts := map[string][]mixedbatchwriter.Option{
+		"defaults":           nil,
+		"batch10k bytes100M": {mixedbatchwriter.WithBatchSizeBytes(100000000), mixedbatchwriter.WithBatchSize(10000)},
+	}
+	streamingbatchwriterOpts := map[string][]streamingbatchwriter.Option{
+		"defaults":  nil,
+		"bytes100M": {streamingbatchwriter.WithBatchSizeBytes(100000000)},
+	}
+	const multiplyN = 1000
+
+	var bCases []bCase
+	bCases = append(bCases, writerMatrix[*batchwriter.BatchWriter, batchwriter.Client, batchwriter.Option]("BatchWriter", batchwriter.New, &batchwriterClient{}, makeRecord, batchwriterOpts)...)
+	bCases = append(bCases, writerMatrix[*batchwriter.BatchWriter, batchwriter.Client, batchwriter.Option]("BatchWriter wide", batchwriter.New, &batchwriterClient{}, makeWideRecord, batchwriterOpts)...)
+	bCases = append(bCases, writerMatrix[*mixedbatchwriter.MixedBatchWriter, mixedbatchwriter.Client, mixedbatchwriter.Option]("MixedBatchWriter", mixedbatchwriter.New, &mixedbatchwriterClient{}, makeRecord, mixedbatchwriterOpts)...)
+	bCases = append(bCases, writerMatrix[*mixedbatchwriter.MixedBatchWriter, mixedbatchwriter.Client, mixedbatchwriter.Option]("MixedBatchWriter wide", mixedbatchwriter.New, &mixedbatchwriterClient{}, makeWideRecord, mixedbatchwriterOpts)...)
+	bCases = append(bCases, writerMatrix[*streamingbatchwriter.StreamingBatchWriter, streamingbatchwriter.Client, streamingbatchwriter.Option]("StreamingBatchWriter", streamingbatchwriter.New, &streamingbatchwriterClient{}, makeRecord, streamingbatchwriterOpts)...)
+	bCases = append(bCases, writerMatrix[*streamingbatchwriter.StreamingBatchWriter, streamingbatchwriter.Client, streamingbatchwriter.Option]("StreamingBatchWriter wide", streamingbatchwriter.New, &streamingbatchwriterClient{}, makeWideRecord, streamingbatchwriterOpts)...)
+
+	for _, c := range bCases {
+		c := c
+		b.Run(c.name, func(b *testing.B) {
+			var (
+				mStart runtime.MemStats
+				mEnd   runtime.MemStats
+			)
+
+			ch := make(chan message.WriteMessage)
+			errCh := make(chan error)
+			go func() {
+				defer close(errCh)
+				errCh <- c.wr.Write(context.Background(), ch)
+			}()
+
+			newN := b.N * multiplyN
+
+			runtime.ReadMemStats(&mStart)
+			b.ResetTimer()
+			for i := 0; i < newN; i++ {
+				rec := c.rec()
+				ch <- &message.WriteInsert{
+					Record: rec,
+				}
+			}
+			close(ch)
+			err := <-errCh
+
+			b.StopTimer()
+
+			if err != nil {
+				b.Fatal(err)
+			}
+
+			runtime.ReadMemStats(&mEnd)
+
+			allocatedBytes := mEnd.Alloc - mStart.Alloc
+			b.ReportMetric(float64(allocatedBytes)/float64(newN), "bytes/op")
```

Comment:
Doesn't Go measure memory usage already if we provide `-benchmem`? What am I missing?

Reply:
`benchmem` will report `allocs/op` (probably the same number as ours but I didn't compare them) and won't report until invoked with `-benchmem` or toggled from within the test. We also multiply the `N` to get more stable benchmarks (and hit writer batching spots).